utils/s3/client.cc

```diff
@@ -223,6 +232,64 @@ future<client::stats> client::get_object_stats(sstring object_name) {
     co_return st;
 }
 
+future<client::TagSet> client::get_object_tagging(sstring object_name) {
+    // see https://docs.aws.amazon.com/AmazonS3/latest/API/API_GetObjectTagging.html
+    auto req = http::request::make("GET", _host, fmt::format("{}?tagging", object_name));
+    s3l.trace("GET {} tagging", object_name);
+    TagSet tagging;
+    co_await make_request(std::move(req),
+                          [&tagging] (const http::reply& reply, input_stream<char>&& in) mutable -> future<> {
+        auto& tagset = tagging;
+        auto input = std::move(in);
+        auto body = co_await util::read_entire_stream_contiguous(input);
+        auto doc = std::make_unique<rapidxml::xml_document<>>();
+        try {
+            doc->parse<0>(body.data());
+        } catch (const rapidxml::parse_error& e) {
+            s3l.warn("cannnot parse tagging response: {}", e.what());
+            throw std::runtime_error("cannot parse tagging response");
+        }
+        auto tagset_node = doc->first_node("Tagging")->first_node("TagSet");
+        for (auto tag_node = tagset_node->first_node("Tag"); tag_node; tag_node = tag_node->next_sibling()) {
+            auto key = tag_node->first_node("Key")->value();
+            auto value = tag_node->first_node("Value")->value();
+            tagset.emplace_back(key, value);
+        }
+    }, http::reply::status_type::ok);
+    co_return tagging;
+}
+
+future<> client::put_object_tagging(sstring object_name, client::TagSet tagging) {
+    // see https://docs.aws.amazon.com/AmazonS3/latest/API/API_PutObjectTagging.html
+    auto req = http::request::make("PUT", _host, fmt::format("{}?tagging", object_name));
+    s3l.warn("PUT {} tagging", object_name);
+    fmt::memory_buffer body;
+    fmt::format_to(fmt::appender(body), "<Tagging><TagSet>{}</TagSet></Tagging>", fmt::join(tagging, ""));
```

Comment:
Please, put both -- size calculation and body-writer -- as a local helper functions next to `prepare_multipart_upload_parts` and `dump_multipart_upload_parts`

Reply:
i extracted the dump logic into a helper. for the same reason explained above i didn't put it next to this helper next to `prepare_multipart_upload_parts()` and `dump_multipart_upload_parts()`.